discoverd/client/client.go

```diff
@@ -36,268 +27,306 @@ func init() {
 }
 
-type Client struct {
-	c *httpclient.Client
+type Config struct {
+	Endpoints []string
+}
 
-	// Logger is used to log messages from Heartbeaters, it must not be changed
-	// after the first API request made with the Client.
-	Logger log15.Logger
+type Client struct {
+	servers map[string]*httpclient.Client
+	pinned  string
+	leader  string
+	idx     uint64
+	mu      sync.RWMutex
+	Logger  log15.Logger
 }
 
-func NewClient() *Client {
-	url := os.Getenv("DISCOVERD")
-	if url == "" {
-		url = "http://127.0.0.1:1111"
+func NewClientWithConfig(config Config) *Client {
+	client := &Client{
+		servers: make(map[string]*httpclient.Client, len(config.Endpoints)),
+		Logger:  defaultLogger,
+	}
+	for _, e := range config.Endpoints {
+		client.servers[e] = client.httpClient(e)
 	}
-	return NewClientWithURL(url)
+	return client
 }
 
 func NewClientWithURL(url string) *Client {
-	if !strings.HasPrefix(url, "http") {
-		url = "http://" + url
-	}
-	return &Client{
-		c: &httpclient.Client{
-			URL: url,
-			HTTP: &http.Client{
-				Transport:     &http.Transport{Dial: dialer.Retry.Dial},
-				CheckRedirect: redirectPreserveHeaders,
-			},
-		},
-		Logger: defaultLogger,
-	}
+	return NewClientWithConfig(Config{Endpoints: formatURLs(strings.Split(url, ","))})
 }
 
-func NewClientWithHTTP(url string, hc *http.Client) *Client {
-	if url == "" {
-		url = "http://127.0.0.1:1111"
+func NewClient() *Client {
+	return NewClientWithConfig(defaultConfig())
+}
+
+func defaultConfig() Config {
+	urls := os.Getenv("DISCOVERD")
+	if urls == "" || urls == "none" {
+		urls = "http://127.0.0.1:1111"
 	}
-	return &Client{
-		c: &httpclient.Client{
-			URL:  url,
-			HTTP: hc,
-		},
-		Logger: defaultLogger,
+	return Config{Endpoints: formatURLs(strings.Split(urls, ","))}
+}
+
+func formatURLs(urls []string) []string {
+	formatted := make([]string, 0, len(urls))
+	for _, u := range urls {
+		if !strings.HasPrefix(u, "http") {
+			u = "http://" + u
+		}
+		formatted = append(formatted, u)
 	}
+	return formatted
 }
 
-func redirectPreserveHeaders(req *http.Request, via []*http.Request) error {
-	if len(via) >= 10 {
-		return fmt.Errorf("too many redirects")
+func (c *Client) updateLeader(host string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for addr, s := range c.servers {
+		if s.Host == host {
+			c.leader = addr
+		}
 	}
-	if len(via) == 0 {
+}
+
+func (c *Client) httpClient(url string) *httpclient.Client {
+	checkRedirect := func(req *http.Request, via []*http.Request) error {
+		if len(via) >= 10 {
+			return fmt.Errorf("too many redirects")
+		}
+		if len(via) > 0 {
+			for attr, val := range via[0].Header {
+				if _, ok := req.Header[attr]; !ok {
+					req.Header[attr] = val
+				}
+			}
+		}
+		c.updateLeader(req.Host)
 		return nil
 	}
-	for attr, val := range via[0].Header {
-		if _, ok := req.Header[attr]; !ok {
-			req.Header[attr] = val
-		}
+	return &httpclient.Client{
+		URL: url,
+		HTTP: &http.Client{
+			CheckRedirect: checkRedirect,
```

Comment:
I don't see why not, I wasn't aware you could share them across httpclient.Client instances but if so  may as well.